File: benchmarks/suite/bench_fit.py
"""`tick` side of the M4 fit benchmarks. See benchmarks/README.md.

Reads the events written by the hawk side, so both fit exactly the same data.
Times the fit call only; loading and construction are outside the clock.

Usage: bench_fit.py <dir> <d> <n,n,...>
"""

import json
import os
import pathlib
import sys
import time
import warnings

import numpy as np

# benchmarks/README.md §2: single-threaded on both sides.
for _v in ("OMP_NUM_THREADS", "OPENBLAS_NUM_THREADS", "MKL_NUM_THREADS",
           "VECLIB_MAXIMUM_THREADS", "NUMEXPR_NUM_THREADS"):
    os.environ[_v] = "1"

from tick.hawkes import HawkesExpKern  # noqa: E402

# §4 and §5, fixed before any number was produced.
WARMUP = 1
TIMED = 5
TRUE_DECAY = 1.0        # §5.1: tick is handed the true beta; it cannot fit it
PENALTY_C = 1e9         # §5.2
TOL = 1e-10             # §5.3
MAX_ITER = 500
SINGLE_RUN_BUDGET = 600.0    # §4.1
CELL_BUDGET = 1800.0         # §4.1


def read_events(path):
    values = path.read_text().split()
    horizon = float(values[0])
    d = int(values[1])
    events, cursor = [], 2
    for _ in range(d):
        count = int(values[cursor]); cursor += 1
        events.append(np.array([float(v) for v in values[cursor:cursor + count]],
                               dtype=np.float64))
        cursor += count
    return horizon, events


def run_one(events, gofit):
    """Warmup then TIMED fits, honouring the §4.1 budget. Returns a record."""
    def make():
        return HawkesExpKern(decays=TRUE_DECAY, gofit=gofit, penalty="l2",
                             C=PENALTY_C, solver="bfgs", tol=TOL,
                             max_iter=MAX_ITER, verbose=False)

    cell_start = time.perf_counter()
    try:
        for _ in range(WARMUP):
            make().fit(events)
    except Exception as exc:                       # §5.4: record, do not retry
        return {"completed": False, "gofit": gofit,
                "abort_reason": f"{type(exc).__name__}: {exc}".strip()}

    elapsed, learner = [], None
    while len(elapsed) < TIMED:
        learner = make()
        start = time.perf_counter()
        try:
            learner.fit(events)
        except Exception as exc:
            return {"completed": False, "gofit": gofit,
                    "abort_reason": f"{type(exc).__name__}: {exc}".strip()}
        seconds = time.perf_counter() - start
        if seconds > SINGLE_RUN_BUDGET:
            return {"completed": False, "gofit": gofit,
                    "abort_reason": "single run exceeded 600 s"}
        elapsed.append(seconds)
        if time.perf_counter() - cell_start > CELL_BUDGET:
            return {"completed": False, "gofit": gofit,
                    "abort_reason": "cell exceeded 1800 s"}

    ordered = sorted(elapsed)
    return {
        "completed": True,
        "gofit": gofit,
        "seconds_median": ordered[TIMED // 2],
        "seconds_min": ordered[0],
        "seconds_max": ordered[-1],
        "seconds_all": elapsed,
        "baseline": [float(v) for v in np.asarray(learner.baseline).ravel()],
        "excitation": [float(v) for v in np.asarray(learner.adjacency).ravel()],
        "decay": TRUE_DECAY,
    }


def main(directory, d, grid):
    directory = pathlib.Path(directory)
    # §5.4: the likelihood objective is only reachable at d = 1. Least-squares is
    # tick's default and the only estimator that runs at d > 1. Both are recorded at
    # d = 1 so the cost of the objective difference is visible where both work.
    objectives = ["likelihood", "least-squares"] if d == 1 else ["least-squares"]

    for nominal in grid:
        horizon, events = read_events(directory / f"events_d{d}_n{nominal}.txt")
        realized = int(sum(len(c) for c in events))
        runs = []
        for gofit in objectives:
            record = {"dimension": d, "nominal_n": nominal, "events": realized,
                      "horizon": horizon}
            record.update(run_one(events, gofit))
            runs.append(record)
            if record["completed"]:
                print(f"d={d} n={nominal} events={realized} gofit={gofit} "
                      f"median={record['seconds_median']:.6f}s", file=sys.stderr)
            else:
                print(f"d={d} n={nominal} gofit={gofit} ABORTED: "
                      f"{record['abort_reason'][:90]}", file=sys.stderr)
        # One file per grid point, written as soon as that point finishes, so a cell
        # killed at the §4.1 budget loses only the cell in flight (see _common.sh).
        body = {"library": "tick", "dimension": d, "warmup": WARMUP, "timed": TIMED,
                "penalty_C": PENALTY_C, "tol": TOL, "max_iter": MAX_ITER,
                "decay_is_input": True, "runs": runs}
        (directory / f"tick_d{d}_n{nominal}.json").write_text(
            json.dumps(body, indent=2) + "\n")


if __name__ == "__main__":
    warnings.filterwarnings("ignore")
    main(sys.argv[1], int(sys.argv[2]), [int(v) for v in sys.argv[3].split(",")])

File: hawkes/examples/bench_fit.rs
//! `hawkes` side of the M4 fit benchmarks. See `benchmarks/README.md`.
//!
//! Generates the shared data for one dimension, writes it out for the `tick` side, and
//! times the fit under the protocol in §4. Enforces the §4.1 budget.
//!
//! Usage: `bench_fit <out_dir> <d> <n,n,...>`

use std::fs;
use std::io::{BufWriter, Write};
use std::time::Instant;

use hawkes::multivariate::{Observation, Parameters, fit, simulate};
use rand::SeedableRng;
use rand_chacha::ChaCha8Rng;

const SEED: u64 = 20_260_831;
const DECAY: f64 = 1.0;
const WARMUP: usize = 1;
const TIMED: usize = 5;
/// §4.1. Same numbers for both libraries.
const SINGLE_RUN_BUDGET: f64 = 600.0;
const CELL_BUDGET: f64 = 1800.0;

/// §3: `mu_i = 0.5/d`, `alpha_ij = 0.6/d`, so the spectral radius is 0.6 at every `d`.
fn truth(d: usize) -> Parameters {
    Parameters::new(vec![0.5 / d as f64; d], vec![0.6 / d as f64; d * d], DECAY)
        .expect("valid parameters")
}

fn write_events(path: &str, horizon: f64, events: &[Vec<f64>]) {
    let file = fs::File::create(path).expect("create events file");
    let mut out = BufWriter::new(file);
    // `{:?}` on f64 is shortest-round-trip, so Python's float() reads back the
    // identical double. Same guarantee the fixture corpus relies on.
    writeln!(out, "{horizon:?}").unwrap();
    writeln!(out, "{}", events.len()).unwrap();
    for component in events {
        writeln!(out, "{}", component.len()).unwrap();
        for t in component {
            writeln!(out, "{t:?}").unwrap();
        }
    }
}

fn json_array(values: &[f64]) -> String {
    let parts: Vec<String> = values.iter().map(|v| format!("{v:?}")).collect();
    format!("[{}]", parts.join(","))
}

/// One file per grid point, written as soon as that point finishes, so a cell killed
/// at the §4.1 budget loses only the cell in flight rather than the whole dimension.
fn write_point(out_dir: &str, d: usize, nominal: usize, truth: &Parameters, run: String) {
    let body = format!(
        r#"{{"library":"hawkes","dimension":{d},"seed":{SEED},"warmup":{WARMUP},"timed":{TIMED},"true_baseline":{},"true_excitation":{},"true_decay":{DECAY:?},"run":{run}}}"#,
        json_array(truth.baseline()),
        json_array(truth.excitation()),
    );
    fs::write(format!("{out_dir}/hawk_d{d}_n{nominal}.json"), body + "\n").unwrap();
}

fn main() {
    let mut args = std::env::args().skip(1);
    let out_dir = args.next().expect("usage: bench_fit <dir> <d> <n,n,...>");
    let d: usize = args.next().expect("dimension").parse().unwrap();
    let grid: Vec<usize> = args
        .next()
        .expect("event counts")
        .split(',')
        .map(|v| v.parse().unwrap())
        .collect();
    fs::create_dir_all(&out_dir).unwrap();

    let truth = truth(d);
    let rate: f64 = truth
        .stationary_mean_intensity()
        .expect("stationary")
        .iter()
        .sum();

    for nominal in grid {
        let horizon = nominal as f64 / rate;
        let mut rng = ChaCha8Rng::seed_from_u64(SEED);
        let events = simulate(&truth, horizon, &mut rng).unwrap();
        let realized: usize = events.iter().map(|c| c.len()).sum();
        write_events(
            &format!("{out_dir}/events_d{d}_n{nominal}.txt"),
            horizon,
            &events,
        );

        let observation = Observation::new(&events, horizon).unwrap();

        let cell_start = Instant::now();
        let mut aborted = None;
        for _ in 0..WARMUP {
            std::hint::black_box(fit(&observation).unwrap());
            if cell_start.elapsed().as_secs_f64() > CELL_BUDGET {
                aborted = Some("cell budget exceeded during warmup");
            }
        }
        let mut elapsed = Vec::with_capacity(TIMED);
        let mut last = None;
        while aborted.is_none() && elapsed.len() < TIMED {
            let start = Instant::now();
            let result = fit(&observation).unwrap();
            let seconds = start.elapsed().as_secs_f64();
            if seconds > SINGLE_RUN_BUDGET {
                aborted = Some("single run exceeded 600 s");
                break;
            }
            elapsed.push(seconds);
            last = Some(result);
            if cell_start.elapsed().as_secs_f64() > CELL_BUDGET {
                aborted = Some("cell exceeded 1800 s");
                break;
            }
        }

        if let Some(reason) = aborted {
            eprintln!("d={d} n={nominal} events={realized} ABORTED: {reason}");
            write_point(
                &out_dir,
                d,
                nominal,
                &truth,
                format!(
                    r#"{{"dimension":{d},"nominal_n":{nominal},"events":{realized},"horizon":{horizon:?},"completed":false,"abort_reason":"{reason}"}}"#
                ),
            );
            continue;
        }

        elapsed.sort_by(|a, b| a.partial_cmp(b).unwrap());
        let result = last.unwrap();
        let spectral = result.parameters.branching_ratio_spectral_radius();
        eprintln!(
            "d={d} n={nominal} events={realized} median={:.6}s nll={:.6} iters={} \
             grad={:.3e} converged={} rho={:.6}",
            elapsed[TIMED / 2],
            result.negative_log_likelihood,
            result.iterations,
            result.gradient_norm,
            result.converged,
            spectral
        );
        write_point(
            &out_dir,
            d,
            nominal,
            &truth,
            format!(
                r#"{{"dimension":{d},"nominal_n":{nominal},"events":{realized},"horizon":{horizon:?},"completed":true,"seconds_median":{:?},"seconds_min":{:?},"seconds_max":{:?},"seconds_all":{},"negative_log_likelihood":{:?},"iterations":{},"gradient_norm":{:?},"converged":{},"objective_evaluations":{},"gradient_evaluations":{},"spectral_radius":{:?},"baseline":{},"excitation":{},"decay":{:?}}}"#,
                elapsed[TIMED / 2],
                elapsed[0],
                elapsed[TIMED - 1],
                json_array(&elapsed),
                result.negative_log_likelihood,
                result.iterations,
                result.gradient_norm,
                result.converged,
                result.objective_evaluations,
                result.gradient_evaluations,
                spectral,
                json_array(result.parameters.baseline()),
                json_array(result.parameters.excitation()),
                result.parameters.decay(),
            ),
        );
    }
}
